README.md
# Papong - A Pong-like NES Game

Welcome to Papong! This is a Pong-like NES game written in C. I made this game following these fantastic tutorials:

* **nesdoug**: https://nesdoug.com


## How to play

* I provide the NES rom already compiled in the file [dist/papong.nes](dist/papong.nes).
* Download a NES emulator. I tested this rom using [Mesen](https://www.mesen.ca/)
* Open the rom with the emulator. You can play with **Up arrow** (up) and **Down arrow** (down) for Player 1. **w** (up) and **s** (down) for player 2.
* The first player to reach **5** points wins.

Here are some images:

![Introduction](docs/images/intro.png)

![Game](docs/images/first.png)

## How to compile it

* Download the [cc65 package](https://cc65.github.io/) which contains the C compiler, assembler and linker for the 6502 (NES).
* Set the **bin** directory of **cc65** in your **PATH** env.
* You can execute the following command to compile the game on Windows:

```bash
.\compile.bat
```

* Or Linux:

```bash
./compile.sh
```

* If everything went OK, then the compiled rom will be available on the **build/papong.nes** file.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- None

## [0.1.2] - 2022-02-26
### Fixed
- Fixed version shown in game.

## [0.1.1] - 2022-02-13
### Fixed
- Minor fixes.

## [0.1.0] - 2022-02-13
### Added
- First version.

crt0.s
; Startup code for cc65 and Shiru's NES library
; based on code by Groepaz/Hitmen <groepaz@gmx.net>, Ullrich von Bassewitz <uz@cc65.org>



FT_BASE_ADR		= $0100		;page in RAM, should be $xx00
FT_DPCM_OFF		= $f000		;$c000..$ffc0, 64-byte steps
FT_SFX_STREAMS	= 1			;number of sound effects played at once, 1..4

FT_THREAD       = 1		;undefine if you call sound effects in the same thread as sound update
FT_PAL_SUPPORT	= 1		;undefine to exclude PAL support
FT_NTSC_SUPPORT	= 1		;undefine to exclude NTSC support
FT_DPCM_ENABLE  = 0		;undefine to exclude all DMC code
FT_SFX_ENABLE   = 1		;undefine to exclude all sound effects code





;REMOVED initlib
;this called the CONDES function

    .export _exit,__STARTUP__:absolute=1
	.import push0,popa,popax,_main,zerobss,copydata

; Linker generated symbols
	.import __STACK_START__   ,__STACKSIZE__ ;changed
	.import __ROM0_START__  ,__ROM0_SIZE__
	.import __STARTUP_LOAD__,__STARTUP_RUN__,__STARTUP_SIZE__
	.import	__CODE_LOAD__   ,__CODE_RUN__   ,__CODE_SIZE__
	.import	__RODATA_LOAD__ ,__RODATA_RUN__ ,__RODATA_SIZE__
	.import NES_MAPPER, NES_PRG_BANKS, NES_CHR_BANKS, NES_MIRRORING

	.importzp _PAD_STATE, _PAD_STATET ;added
    .include "zeropage.inc"






PPU_CTRL	=$2000
PPU_MASK	=$2001
PPU_STATUS	=$2002
PPU_OAM_ADDR=$2003
PPU_OAM_DATA=$2004
PPU_SCROLL	=$2005
PPU_ADDR	=$2006
PPU_DATA	=$2007
PPU_OAM_DMA	=$4014
PPU_FRAMECNT=$4017
DMC_FREQ	=$4010
CTRL_PORT1	=$4016
CTRL_PORT2	=$4017

OAM_BUF		=$0200
PAL_BUF		=$01c0
VRAM_BUF	=$0700



.segment "ZEROPAGE"

NTSC_MODE: 			.res 1
FRAME_CNT1: 		.res 1
FRAME_CNT2: 		.res 1
VRAM_UPDATE: 		.res 1
NAME_UPD_ADR: 		.res 2
NAME_UPD_ENABLE: 	.res 1
PAL_UPDATE: 		.res 1
PAL_BG_PTR: 		.res 2
PAL_SPR_PTR: 		.res 2
SCROLL_X: 			.res 1
SCROLL_Y: 			.res 1
SCROLL_X1: 			.res 1
SCROLL_Y1: 			.res 1
PAD_STATE: 			.res 2		;one byte per controller
PAD_STATEP: 		.res 2
PAD_STATET: 		.res 2
PPU_CTRL_VAR: 		.res 1
PPU_CTRL_VAR1: 		.res 1
PPU_MASK_VAR: 		.res 1
RAND_SEED: 			.res 2
FT_TEMP: 			.res 3

TEMP: 				.res 11
SPRID:				.res 1

PAD_BUF		=TEMP+1

PTR			=TEMP	;word
LEN			=TEMP+2	;word
NEXTSPR		=TEMP+4
SCRX		=TEMP+5
SCRY		=TEMP+6
SRC			=TEMP+7	;word
DST			=TEMP+9	;word

RLE_LOW		=TEMP
RLE_HIGH	=TEMP+1
RLE_TAG		=TEMP+2
RLE_BYTE	=TEMP+3

;nesdoug code requires
VRAM_INDEX:			.res 1
META_PTR:			.res 2
DATA_PTR:			.res 2





.segment "HEADER"

    .byte $4e,$45,$53,$1a
	.byte <NES_PRG_BANKS
	.byte <NES_CHR_BANKS
	.byte <NES_MIRRORING|(<NES_MAPPER<<4)
	.byte <NES_MAPPER&$f0
	.res 8,0



.segment "STARTUP"

start:
_exit:

    sei
	cld
	ldx #$40
	stx CTRL_PORT2
    ldx #$ff
    txs
    inx
    stx PPU_MASK
    stx DMC_FREQ
    stx PPU_CTRL		;no NMI

initPPU:
    bit PPU_STATUS
@1:
    bit PPU_STATUS
    bpl @1
@2:
    bit PPU_STATUS
    bpl @2

clearPalette:
	lda #$3f
	sta PPU_ADDR
	stx PPU_ADDR
	lda #$0f
	ldx #$20
@1:
	sta PPU_DATA
	dex
	bne @1

clearVRAM:
	txa
	ldy #$20
	sty PPU_ADDR
	sta PPU_ADDR
	ldy #$10
@1:
	sta PPU_DATA
	inx
	bne @1
	dey
	bne @1

clearRAM:
    txa
@1:
    sta $000,x
    sta $100,x
    sta $200,x
    sta $300,x
    sta $400,x
    sta $500,x
    sta $600,x
    sta $700,x
    inx
    bne @1

	lda #4
	jsr _pal_bright
	jsr _pal_clear
	jsr _oam_clear

    jsr	zerobss
	jsr	copydata

    lda #<(__STACK_START__+__STACKSIZE__) ;changed
    sta	sp
    lda	#>(__STACK_START__+__STACKSIZE__)
    sta	sp+1            ; Set argument stack ptr

;	jsr	initlib
; removed. this called the CONDES function

	lda #%10000000
	sta <PPU_CTRL_VAR
	sta PPU_CTRL		;enable NMI
	lda #%00000110
	sta <PPU_MASK_VAR

waitSync3:
	lda <FRAME_CNT1
@1:
	cmp <FRAME_CNT1
	beq @1

detectNTSC:
	ldx #52				;blargg's code
	ldy #24
@1:
	dex
	bne @1
	dey
	bne @1

	lda PPU_STATUS
	and #$80
	sta <NTSC_MODE

	jsr _ppu_off

	lda #0
	ldx #0
	jsr _set_vram_update

	ldx #<music_data
	ldy #>music_data
	lda <NTSC_MODE
	jsr FamiToneInit

	.if(FT_SFX_ENABLE)
	ldx #<sounds_data
	ldy #>sounds_data
	jsr FamiToneSfxInit
	.endif

	lda #$fd
	sta <RAND_SEED
	sta <RAND_SEED+1

	lda #0
	sta PPU_SCROLL
	sta PPU_SCROLL

	jmp _main			;no parameters

	.include "lib/neslib.s"
	.include "lib/nesdoug.s"
	.include "music/famitone2.s"
	
	

.segment "RODATA"

music_data:
;	.include "music.s"



	.if(FT_SFX_ENABLE)
sounds_data:
;	.include "sounds.s"
	.endif

	
	
.segment "SAMPLES"
;	.incbin "music_dpcm.bin"



.segment "VECTORS"

    .word nmi	;$fffa vblank nmi
    .word start	;$fffc reset
   	.word irq	;$fffe irq / brk


.segment "CHARS"

	.incbin "Alpha.chr"

papong.h
const unsigned char BLACK=0x00;
const unsigned char LEFT_LIMIT=16;
const unsigned char RIGHT_LIMIT=240;
const unsigned char TOP_LIMIT = 16;
const unsigned char BOTTOM_LIMIT=214;
const unsigned char MOVEMENT_NONE = 0;
const unsigned char MOVEMENT_UP = 1;
const unsigned char MOVEMENT_DOWN = 2;
const unsigned char MOVEMENT_LEFT = 3;
const unsigned char MOVEMENT_RIGHT = 4;
const unsigned char PADDLE_VELOCITY = 2;
const unsigned char BALL_INITIAL_X = 120;
const unsigned char BALL_INITIAL_Y = 120;
const unsigned char BALL_WIDTH = 8;
const unsigned char BALL_HEIGHT = 8;
const unsigned char PLAYER_INITIAL_Y = 80;
const unsigned char PLAYER_WIDTH = 8;
const unsigned char PLAYER_HEIGHT = 32;
const unsigned char PLAYER_1_INITIAL_X = 30;
const unsigned char PLAYER_2_INITIAL_X = 218;
const unsigned char MAX_POINTS = 5;

const unsigned char PlayerSprite[]={
    0,  0,0x00,1,
    0, 16,0x00,1,
    0,  8,0x00,1,
    0, 24,0x00,1,
    128
};

const unsigned char BallSprite[]={
    0,  0,0x01,3,
    128
};

const unsigned char Background[]={
    0x01,0x02,0x02,0x50,0x4c,0x41,0x59,0x45,0x52,0x20,0x31,0x3a,0x10,0x01,0x02,0x02,
    0x01,0x03,0x50,0x4c,0x41,0x59,0x45,0x52,0x20,0x32,0x3a,0x10,0x01,0x02,0x02,0x01,
    0x23,0x20,0x01,0x1b,0x02,0x01,0x03,0x20,0x01,0x1b,0x02,0x01,0x03,0x20,0x01,0x1b,
    0x02,0x01,0x03,0x20,0x01,0x1b,0x02,0x01,0x03,0x20,0x01,0x1b,0x02,0x01,0x03,0x20,
    0x01,0x1b,0x02,0x01,0x03,0x20,0x01,0x1b,0x02,0x01,0x03,0x20,0x01,0x1b,0x02,0x01,
    0x03,0x20,0x01,0x1b,0x02,0x01,0x03,0x20,0x01,0x1b,0x02,0x01,0x03,0x20,0x01,0x1b,
    0x02,0x01,0x03,0x20,0x01,0x1b,0x02,0x01,0x03,0x20,0x01,0x1b,0x02,0x01,0x03,0x20,
    0x01,0x1b,0x02,0x01,0x03,0x20,0x01,0x1b,0x02,0x01,0x03,0x20,0x01,0x16,0x10,0x20,
    0x01,0x03,0x02,0x01,0x03,0x20,0x01,0x1b,0x02,0x01,0x03,0x20,0x01,0x1b,0x02,0x01,
    0x03,0x20,0x01,0x1b,0x02,0x01,0x03,0x20,0x01,0x1b,0x02,0x01,0x03,0x20,0x01,0x1b,
    0x02,0x01,0x03,0x20,0x01,0x1b,0x02,0x01,0x03,0x20,0x01,0x1b,0x02,0x01,0x03,0x20,
    0x01,0x1b,0x02,0x01,0x03,0x20,0x01,0x0e,0x10,0x01,0x08,0x20,0x01,0x03,0x02,0x01,
    0x39,0x56,0x20,0x01,0x04,0x02,0x01,0x21,0x00,0x01,0x3e,0x00,0x01,0x00
};

const unsigned char palette_bg[]={
    0x0f, 0x00, 0x10, 0x30, // black, gray, lt gray, white
    0,0,0,0,
    0,0,0,0,
    0,0,0,0
};

const unsigned char palette_spr[]={
    0x0f,0x00,0x10,0x30,
    0x0f,0x11,0x21,0x31,
    0x0f,0x06,0x16,0x26,
    0x0f,0x09,0x19,0x29
};

papong.c
#include "lib/neslib.h"
#include "lib/nesdoug.h"
#include "papong.h"

#pragma bss-name(push, "ZEROPAGE")

// Some structs

struct Ball {
	unsigned char x;
	unsigned char y;
	unsigned char width;
	unsigned char height;
	unsigned char x_direction;
	unsigned char y_direction;
};

struct Player {
	unsigned char x;
	unsigned char y;
	unsigned char width;
	unsigned char height;
	unsigned char y_direction;
	unsigned char pad;
	unsigned char pad_number;
	unsigned char points;
	const unsigned char player_win_text[15];
};

struct Ball ball = {120, 120, 8, 8, 4, 2};
struct Player player1 = {30, 80, 8, 32, 0, 0, 0, 0, "Player 1 WINS!"};
struct Player player2 = {218, 80, 8, 32, 0, 0, 1, 0, "Player 2 WINS!"};

// Functions

void fill_background(const unsigned char color) {
	ppu_off();

	vram_fill(color, NTADR_A(0, 0));

	ppu_on_all();
}

void clear_background(void) {
	fill_background(BLACK);
}

void show_text(const unsigned char * text, int x, int y) {
	static unsigned char i = 0;

	ppu_off();

	pal_bg(palette_bg);

	vram_adr(NTADR_A(x, y));

	i = 0;

	while(text[i]) {
		vram_put(text[i]);

		++i;
	}

	ppu_on_all();
}

void show_introduction(void) {
	static const char first_text[] = "Molleja Soft presents...";
	static const char second_text[] = "PAPONG";
    static const char clear_text[] = "                        ";
	static const char version_text[] = "0.1.2";

    show_text(version_text, 25, 28);

	show_text(first_text, 4, 12);

	delay(100);

    show_text(clear_text, 4, 12);

	show_text(second_text, 13, 12);

	delay(100);

    show_text(clear_text, 4, 12);
}

void initialize_system(void) {
	// Screen off

	ppu_off();

	// Load the BG palette

	pal_bg(palette_bg);

	// Load the Sprites palette

	pal_spr(palette_spr);

	// Use the second set of tiles for the sprites

	bank_spr(1);

    // Create buffer and clear it

    set_vram_buffer();
    clear_vram_buffer();

    // Draw background

    vram_adr(NAMETABLE_A);
    vram_unrle(Background);

	// Turn on screen

	ppu_on_all();
}

void restart_round() {
	player1.y = PLAYER_INITIAL_Y;
	player2.y = PLAYER_INITIAL_Y;
	ball.x = BALL_INITIAL_X;
	ball.y = BALL_INITIAL_Y;
}

void add_point_to_player(struct Player * player) {
	++(player->points);

	if (player->points >= MAX_POINTS) {
		show_text(player->player_win_text, 9, 12);

		delay(100);

		show_text("              ", 9, 12);

		player1.points = 0;
		player2.points = 0;
	}

	delay(100);

	restart_round();
}

void ball_movement(void){
	long collisionPlayer1 = 0;
	long collisionPlayer2 = 0;

	if (ball.x + ball.width >= RIGHT_LIMIT) {
		add_point_to_player(&player1);

		return;
	}

	if (ball.x <= LEFT_LIMIT) {
		add_point_to_player(&player2);

		return;
	}

	collisionPlayer1 = check_collision(&ball, &player1);

	if (!collisionPlayer1) {
		collisionPlayer2 = check_collision(&ball, &player2);
	}

	if (collisionPlayer2) {
		ball.x_direction = MOVEMENT_LEFT;
	} else if (collisionPlayer1) {
		ball.x_direction = MOVEMENT_RIGHT;
	}

	if (ball.y + ball.height >= BOTTOM_LIMIT) {
		ball.y_direction = MOVEMENT_UP;
	} else if (ball.y <= TOP_LIMIT) {
		ball.y_direction = MOVEMENT_DOWN;
	}

	if (ball.x_direction == MOVEMENT_LEFT) {
		ball.x -= 1;
	} else {
		ball.x += 1;
	}

	if (ball.y_direction == MOVEMENT_UP) {
		ball.y -= 1;
	} else {
		ball.y += 1;
	}
}

void player_movement(struct Player * player) {
	// Read the controller

	player->pad = pad_poll(player->pad_number);

	// Determine Player movement

	if (player->pad & PAD_UP){
		player->y_direction = MOVEMENT_UP;
	} else if (player->pad & PAD_DOWN){
		player->y_direction = MOVEMENT_DOWN;
	} else {
		return;
	}

	// Check screen limits

	if (player->y_direction == MOVEMENT_UP && player->y <= TOP_LIMIT) {
		return;
	}

	if (player->y_direction == MOVEMENT_DOWN && player->y + player->height >= BOTTOM_LIMIT) {
		return;
	}

	// Move paddle

	if (player->y_direction == MOVEMENT_UP) {
		player->y -= PADDLE_VELOCITY;
	} else {
		player->y += PADDLE_VELOCITY;
	}
}

void draw(void){
    // Clear vram buffer

    clear_vram_buffer();

	// Clear all sprites from sprite buffer

	oam_clear();

	// Draw points

	// NOTE: Here we add 48 to match the # Tile of the corresponding number with the score of the player

	one_vram_buffer(player1.points + 48, NTADR_A(13, 0));
	one_vram_buffer(player2.points + 48, NTADR_A(29, 0));

	// Draw our ball

	oam_meta_spr(ball.x, ball.y, BallSprite);

	// Draw our first player

	oam_meta_spr(player1.x, player1.y, PlayerSprite);

	// Draw our second player

	oam_meta_spr(player2.x, player2.y, PlayerSprite);
}

void main(void) {
	initialize_system();

	show_introduction();

	while (1) {
		// :: Game Loop

		// Wait till beginning of the frame

		ppu_wait_nmi();

		// Determine players movements

		player_movement(&player1);
		player_movement(&player2);

		// Determine ball movement

		ball_movement();

		// Draw everything!

        draw();
	}
}